### ISSUE TITLE
Cannot redefine/reassign a property in the same file

### ISSUE THREAD

**Issue by BinToss** [tintoy/msbuild-project-tools-server#100]
`System.ArgumentException: An item with the same key has already been added.`
`System.InvalidOperationException: Parent project does not have an MSBuild project.`

My ZipPublishDir Target defines a `Destination` property. Then it conditionally appends to the value.
```xml
<RepoRoot>$([System.IO.Path]::GetDirectoryName($([MSBuild]::GetPathOfFileAbove('.gitignore', '$(MSBuildProjectDirectory)'))))</RepoRoot>
<Destination>$(RepoRoot)/publish/$(AssemblyName).$(Version)$(PART_RuntimeIdentifier)</Destination>
<Destination Condition="'$(RuntimeIdentifier)' != '' ">$(Destination).$(RuntimeIdentifier)</Destination>
```

MSBuild reassigns the value to the existing property–or redefines it–as expected.
`msbuild-project-tools-server` cannot handle reassignments. Instead, it throws `System.ArgumentException`. Consequently, it cannot load any project importing a file containing this usage.

When adding a key that already exists, `msbuild-project-tools-server` should
- **A**: Assign the new key's value to the existing key. All instances of that property will have the same evaluated value.
- **B**: Keep both, but rename duplicates in the order they're discovered. This allows all instances to evaluate to independent values e.g. if the first definition is `Destination`, then the second definition's key would be `Destination>1` and so on.
  - The use of the `>` character should prevent naming collisions. It cannot be used in a property name—even as `&gt;`.

----

**error caused by ZipPublishDir**
```log
[Error - 5:23:01 AM] Unhandled exception in OnHover.
System.ArgumentException: An item with the same key has already been added. Key: file:///c%3A/Repos/BinToss/GroupBox.Avalonia/node_modules/%40halospv3/hce.shared-config/dotnet/ZipPublishDir.targets
   at bool System.Collections.Generic.Dictionary<TKey, TValue>.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at Task<ProjectDocument> MSBuildProjectTools.LanguageServer.Documents.Workspace.GetProjectDocument(Uri documentUri, bool reload)+(Uri _) => { }
   at async Task<ProjectDocument> MSBuildProjectTools.LanguageServer.Documents.Workspace.GetProjectDocument(Uri documentUri, bool reload)
   at async Task<Hover> MSBuildProjectTools.LanguageServer.Handlers.HoverHandler.OnHover(TextDocumentPositionParams parameters, CancellationToken cancellationToken)
   at async Task<Hover> MSBuildProjectTools.LanguageServer.Handlers.HoverHandler.OmniSharp.Extensions.JsonRpc.IRequestHandler<OmniSharp.Extensions.LanguageServer.Protocol.Models.TextDocumentPositionParams,OmniSharp.Extensions.LanguageServer.Protocol.Models.Hover>.Handle(?)

```
**ZipPublishDir**
```csproj
<Project>
  <!-- https://learn.microsoft.com/en-us/visualstudio/msbuild/zipdirectory-task?view=vs-2022 -->
  <Target Name="ZipPublishDir" AfterTargets="Publish">
    <PropertyGroup>
      <RepoRoot>$([System.IO.Path]::GetDirectoryName($([MSBuild]::GetPathOfFileAbove('.gitignore', '$(MSBuildProjectDirectory)'))))</RepoRoot>
      <Destination>$(RepoRoot)/publish/$(AssemblyName).$(Version)$(PART_RuntimeIdentifier)</Destination>
      <Destination Condition="'$(RuntimeIdentifier)' != '' ">$(Destination).$(RuntimeIdentifier)</Destination>
    </PropertyGroup>

    <ZipDirectory
      SourceDirectory="$(PublishDir)"
      DestinationFile="$(Destination).zip"
      Overwrite="true" />
  </Target>
</Project>
```

**error in project importing ZipPublishDir**
```log
[Error - 5:36:55 AM] Error loading MSBuild project '"c:\Repos\BinToss\GroupBox.Avalonia\GroupBox.Avalonia.Sample\GroupBox.Avalonia.Sample.csproj"'.
System.InvalidOperationException: Parent project does not have an MSBuild project.
   at bool MSBuildProjectTools.LanguageServer.Documents.SubProjectDocument.TryLoadMSBuildProject()
[Warn  - 5:36:55 AM] Reloaded project file "c:\Repos\BinToss\GroupBox.Avalonia\GroupBox.Avalonia.Sample\GroupBox.Avalonia.Sample.csproj" (XML is valid, but MSBuild project is not).
```

**Comment by tintoy** [comment]
Thanks for the detailed report! I’ll look into this tomorrow morning 🙂

**Comment by tintoy** [comment]
Sorry for the delayed response (production issues at work 😂). I think I can see a solution to this problem, I just need to try it out, and will let you know when there is something to test.

**Comment by tintoy** [comment]
(part of the problem is that MSBuild has 3 different semantic models (or layers) that can be used to represent various aspects of the project. MSBuild’a heavy lifting is largely concerned with mapping between the first 2 of these models/layers: Construction (project XML) and Evaluation (evaluated property values and item groups). Part of the issue is that the mapping is not (and cannot always be) one-to-one; the construction project has 2 “Destination” property elements but the evaluation project only has one “Destination” property value.

The original assumption was that we would only track the last declaration of a property in a project and use that one to link it to the actual property in the evaluation project:

https://github.com/tintoy/msbuild-project-tools-server/blob/9f85d757ca7a324163129a90aac09b534a3723d3/src/LanguageServer.SemanticModel.MSBuild/MSBuildProperty.cs#L25

We have done a fair bit of refactoring lately and I suspect a bug has crept in somewhere that is failing to account for multiple declarations.

**Comment by tintoy** [comment]
Ok, I've at least managed to validate that the language model does still correctly support properties defined and then redefined in the same project file. So it must be an upstream component in the language server. Will keep digging :)

**Comment by tintoy** [comment]
I'm starting to think this may actually be an issue with the targets file being picked up more than once during parsing.

**Comment by tintoy** [comment]
I think 0d40200dc1f7e9685c277818eae2ab9fb7f97847 may wind up fixing the issue.

**Comment by BinToss** [comment]
I still haven't gotten around to building the potential fix and checking if it solves the issue. I'll see if I can do it later today.

**Comment by tintoy** [comment]
Thanks! I’ve been away for the last couple of days but I’ll try to get a package built tomorrow.

**Comment by tintoy** [comment]
@BinToss - can you give this package 
[msbuild-project-tools-0.6.3-redefine-property.2.zip](https://github.com/tintoy/msbuild-project-tools-server/files/14817952/msbuild-project-tools-0.6.3-redefine-property.2.zip)
a try?

**Comment by BinToss** [comment]
Give the following PropertyGroup of an imported file...
```xml
<!-- ZipPublishDir.targets -->
<Project>
	<PropertyGroup>
		<!-- 
			Unused value: false
			Actual value: true
		-->
		<ANewVariable>false</ANewVariable>
		<!-- Value: true -->
		<ANewVariable>true</ANewVariable>
  	</PropertyGroup>

	<Target ...
```
...the IntelliSense value of `ANewVariable` is evaluated correctly and without error in consumer projects.
```xml
	<!-- Condition Evaluated: true, Value: true -->
	<ANewVariable Condition="$(ANewVariable)">$(ANewVariable)</ANewVariable>
```


Given this reproducible example, the evaluation issue is not present in that package!

**Comment by tintoy** [comment]
Great - thanks for verifying! I’ll publish a new version as soon as I’ve addressed some minor PR feedback 🙂

**Comment by tintoy** [comment]
I've published v0.6.3 of the extension to the VS gallery (it may take a couple of minutes to show up as an update in VS Code).

Thanks again for contributing!
